02_knight_and_princess/danil_dyachenko/graph_gen.cpp

```diff
@@ -0,0 +1,122 @@
+#include <array>
+#include <fstream>
+#include <iostream>
+#include <vector>
+
+using VertexId = int;
+using EdgeId = int;
+
+struct Vertex {
+  VertexId id;
+  std::vector<int> edges;
+
+  std::string& to_json() {
+    static std::string string;
+    string.clear();
+    string += "{ \"id\": ";
+    string += std::to_string(id);
+    string += ", \"edge_ids\": [";
+    for (auto it = edges.begin(); it != edges.end() - 1; it++) {
+      string += std::to_string(*it);
+      string += ", ";
+    }
+    string += std::to_string(*(edges.end() - 1));
+    string += "] }";
+    return string;
+  }
+};
+
+struct Edge {
+  EdgeId id;
+  std::array<int, 2> vertices;
+
+  std::string& to_json() {
+    static std::string string;
+    string.clear();
+    string += "{ \"id\": ";
+    string += std::to_string(id);
+    string += ", \"vertex_ids\": [";
+    string += std::to_string(vertices[0]);
+    string += ", ";
+    string += std::to_string(vertices[1]);
+    string += "] }";
+    return string;
+  }
+};
+
+class Graph {
+ public:
+  Graph(const std::vector<Vertex>& vertices, const std::vector<Edge>& edges)
+      : vertices_(vertices), edges_(edges) {}
```

Comment:
Усложняешь себе жизнь с таким конструктором. Лучше бы Graph сам отвечал за добавление вершин и граней, можно было бы избежать сложной валидирующей логики, но хозяин - барин)

В таком случае, тебе нужно удостовериться что входные параметры имеют смысл. Соотвественно грани должны соединять вершины, которые сюда передаются. Иначе они могут соединять несуществующие вершины. И второе, все вершины должны быть соединены, нужно проверить что тут нет вершин без граней.
```suggestion
    Graph(const std::vector<Vertex>& vertices, const std::vector<Edge>& edges)
      : vertices_(vertices), edges_(edges) {
      // validation  
    }
```